server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -230,46 +231,71 @@ private class GetSnapshotsOperation {
+                            if (skipRepository(repoName)) {
+                                continue;
+                            }
 
-                    getSingleRepoSnapshotInfo(repoName, listeners.acquire((SnapshotsInRepo snapshotsInRepo) -> {
-                        allSnapshotInfos.add(snapshotsInRepo.snapshotInfos());
-                        remaining.addAndGet(snapshotsInRepo.remaining());
-                        totalCount.addAndGet(snapshotsInRepo.totalCount());
-                    }).delegateResponse((l, e) -> {
-                        if (isMultiRepoRequest && e instanceof ElasticsearchException elasticsearchException) {
-                            failuresByRepository.put(repoName, elasticsearchException);
-                            l.onResponse(SnapshotsInRepo.EMPTY);
-                        } else {
-                            l.onFailure(e);
+                            SubscribableListener
+
+                                .<RepositoryData>newForked(repositoryDataListener -> {
+                                    if (snapshotNamePredicate == SnapshotNamePredicate.MATCH_CURRENT_ONLY) {
+                                        repositoryDataListener.onResponse(null);
+                                    } else {
+                                        repositoriesService.repository(repoName).getRepositoryData(executor, repositoryDataListener);
+                                    }
+                                })
+
+                                .<SnapshotsInRepo>andThen((l, repositoryData) -> loadSnapshotInfos(repoName, repositoryData, l))
+
+                                .addListener(new DelegatingActionListener<>(listeners.acquire()) {
+                                    @Override
+                                    public void onResponse(SnapshotsInRepo snapshotsInRepo) {
+                                        allSnapshotInfos.add(snapshotsInRepo.snapshotInfos());
+                                        remaining.addAndGet(snapshotsInRepo.remaining());
+                                        totalCount.addAndGet(snapshotsInRepo.totalCount());
+                                        delegate.onResponse(null);
+                                    }
+
+                                    @Override
+                                    public void onFailure(Exception e) {
+                                        if (isMultiRepoRequest && e instanceof ElasticsearchException elasticsearchException) {
+                                            failuresByRepository.put(repoName, elasticsearchException);
+                                            delegate.onResponse(null);
+                                        } else {
+                                            delegate.onFailure(e);
+                                        }
+                                    }
+                                }, executor, threadPool.getThreadContext());
```

Comment:
`loadSnapshotInfos` would also typically complete its listener on `SNAPSHOT_META` rather than the calling threadpool - adding `executor` here ensures that we fork back to `MANAGEMENT` before proceeding